src/QsCompiler/Transformations/TypeResolutionSummary.cs

```diff
@@ -0,0 +1,390 @@
+               .ToLookup(// Reverse the keys and resulting type parameters to make the lookup.
+                   kvp => kvp.value,
+                   kvp => kvp.Key);
+        }
+
+        // Fields and Properties
+
+        /// <summary>
+        /// Array of all the type parameter resolution dictionaries that are summarized in this summary.
+        /// </summary>
+        public readonly ImmutableArray<TypeParameterResolutions> ResolutionsInSummary;
+
+        /// <summary>
+        /// The resulting resolution dictionary from combining all the input resolutions in
+        /// ResolutionsInSummary. Represents a summary of all the type parameters found and their
+        /// ultimate type resolutions.
+        /// </summary>
+        public TypeParameterResolutions CombinedTypeParameterResolutions { get; private set; }
+
+        /// <summary>
+        /// Flag for if there were any invalid scenarios encountered while creating the summary.
+        /// Invalid scenarios include a type parameter being assigned to multiple conflicting types
+        /// and a type parameter being assigned to a type referencing a different type parameter of
+        /// the same callable. Has value true if no invalid scenarios were encountered.
+        /// </summary>
+        public bool IsValidSummary { get => !this.SummarizesOverConflictingResolution && !this.SummarizesOverParameterConstriction; }
+
+        /// <summary>
+        /// Flag for if, at any time in the creation of the summary, there was a type parameter that
+        /// was assigned conflicting type resolutions. Has value true if a conflict was encountered.
+        /// </summary>
+        public bool SummarizesOverConflictingResolution { get; private set; } = false;
+
+        /// <summary>
+        /// Flag for if, at any time in the creation of the summary, there was a type parameter that
+        /// was assigned a type resolution referencing a different type parameter of the same callable.
+        /// </summary>
+        public bool SummarizesOverParameterConstriction { get; private set; } = false;
+
+        // Constructors
+
+        /// <summary>
+        /// Creates a type parameter resolution summary from the independent type parameter resolutions
+        /// found in the given typed expression and its sub expressions. Only sub-expressions whose
+        /// type parameter resolutions are relevant to the given expression's type parameter resolutions
+        /// are considered.
+        /// </summary>
+        public TypeResolutionSummary(TypedExpression expression) : this(GetTypeParameterResolutions.Apply(expression))
+        {
+        }
+
+        /// <summary>
+        /// Creates a type parameter resolution summary from independent type parameter resolution dictionaries.
+        /// The given resolutions are expected to be ordered such that dictionaries containing type parameters resolutions that
```

Comment:
Can this be enforced? Maybe iterating over the list of dictionaries and if a dependency is out of order raise an exception?

Reply:
This is enforced through the public constructor, which is the preferred way to construct summaries since it can automatically get all the relevant resolutions in the correct order. This other constructor is used mostly for the Call Graph logic. The way to tell if the order is correct is by checking the validation flags, rather than by throwing an exception.

Reply:
Thanks for the explanation!

One more question regarding this constructor. Is there a reason to choose `internal` instead of `private` in this case?

Reply:
Just that the call graph logic uses it.